### ISSUE TITLE
Installation with Python 3.5 no longer working

### ISSUE THREAD

**Issue by eljasala** [oracle/oci-cli#365]
Latest version of pip uses f-strings which were added on Python 3.6. As the install script automatically updates pip to the latest version, this causes installation to fail.

```
-- Executing: ['/home/ubuntu/lib/oracle-cli/bin/pip', 'install', '--upgrade', 'pip']
Collecting pip
  Downloading https://files.pythonhosted.org/packages/de/47/58b9f3e6f611dfd17fb8bd9ed3e6f93b7ee662fb85bdfee3565e8979ddf7/pip-21.0-py3-none-any.whl (1.5MB)
    100% |████████████████████████████████| 1.5MB 1.3MB/s
Installing collected packages: pip
  Found existing installation: pip 8.1.1
    Uninstalling pip-8.1.1:
      Successfully uninstalled pip-8.1.1
Successfully installed pip-21.0
-- Executing: ['/home/ubuntu/lib/oracle-cli/bin/pip', 'install', '--cache-dir', '/tmp/tmpn73vrb04', 'wheel', '--upgrade']
Traceback (most recent call last):
  File "/home/ubuntu/lib/oracle-cli/bin/pip", line 7, in <module>
    from pip._internal.cli.main import main
  File "/home/ubuntu/lib/oracle-cli/lib/python3.5/site-packages/pip/_internal/cli/main.py", line 60
    sys.stderr.write(f"ERROR: {exc}")
                                   ^
SyntaxError: invalid syntax
-- System was unable to use venv, is going to download and create virtualenv.
...
-- Executing: ['/usr/bin/python3', 'virtualenv.py', '--python', '/usr/bin/python3', '/home/ubuntu/lib/oracle-cli']
Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in /home/ubuntu/lib/oracle-cli/bin/python3
Traceback (most recent call last):
  File "virtualenv.py", line 2311, in <module>
    main()
  File "virtualenv.py", line 703, in main
    symlink=options.symlink)
  File "virtualenv.py", line 916, in create_environment
    site_packages=site_packages, clear=clear, symlink=symlink))
  File "virtualenv.py", line 1209, in install_python
    shutil.copyfile(executable, py_executable)
  File "/usr/lib/python3.5/shutil.py", line 98, in copyfile
    raise SameFileError("{!r} and {!r} are the same file".format(src, dst))
shutil.SameFileError: '/usr/bin/python3' and '/home/ubuntu/lib/oracle-cli/bin/python3' are the same file
Traceback (most recent call last):
  File "/tmp/oci_cli_install_tmp_vXXK", line 254, in create_virtualenv
    upgrade_pip_wheel(tmp_dir, install_dir)
  File "/tmp/oci_cli_install_tmp_vXXK", line 236, in upgrade_pip_wheel
    exec_command(cmd)
  File "/tmp/oci_cli_install_tmp_vXXK", line 141, in exec_command
    subprocess.check_call(command_list, cwd=cwd, env=env)
  File "/usr/lib/python3.5/subprocess.py", line 581, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/home/ubuntu/lib/oracle-cli/bin/pip', 'install', '--cache-dir', '/tmp/tmpn73vrb04', 'wheel', '--upgrade']' returned non-zero exit status 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/oci_cli_install_tmp_vXXK", line 685, in <module>
    main()
  File "/tmp/oci_cli_install_tmp_vXXK", line 637, in main
    create_virtualenv(tmp_dir, install_dir)
  File "/tmp/oci_cli_install_tmp_vXXK", line 257, in create_virtualenv
    download_and_create_virtualenv(tmp_dir, install_dir)
  File "/tmp/oci_cli_install_tmp_vXXK", line 202, in download_and_create_virtualenv
    exec_command(cmd, cwd=working_dir)
  File "/tmp/oci_cli_install_tmp_vXXK", line 141, in exec_command
    subprocess.check_call(command_list, cwd=cwd, env=env)
  File "/usr/lib/python3.5/subprocess.py", line 581, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3', 'virtualenv.py', '--python', '/usr/bin/python3', '/home/ubuntu/lib/oracle-cli']' returned non-zero exit status 1
```

**Comment by kernleee** [comment]
Hi, @eljasala
Thank you for reporting this issue.
Pip has dropped support for Python 2.7/3.5 as of their latest release 21.0.
We are aware of this issue and will work on a temporary fix to be released in a week.
To get around this issue at this time please install Python 3.6+ as we are also planning on dropping Python 2.7/3.5 support in the near future.

**Comment by kernleee** [comment]
@eljasala 
I tried the installer script on Python 3.5.3 and it installed successfully. Can you tell me which python version you are using. In my case, pip detected I had 3.5 and did not allow me to install pip==21.0 instead opting for 20.3.4.

**Comment by eljasala** [comment]
Hey @kernleee

Thanks for response. I tried it with Python 3.5.2 which is the version included in Ubuntu 16.04. The image I used was https://docs.oracle.com/en-us/iaas/images/image/eb4c756e-a8b9-447c-ab98-36f116789a44/. I just confirmed that it happens with both fresh image & one where I did `apt-get update && apt-get upgrade`.

Following is full log from fresh image:
```
$ bash -c "$(curl -L https://raw.githubusercontent.com/oracle/oci-cli/master/scripts/install/install.sh)"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 16053  100 16053    0     0   155k      0 --:--:-- --:--:-- --:--:--  156k

    ******************************************************************************
    You have started the OCI CLI Installer in interactive mode. If you do not wish
    to run this in interactive mode, please include the --accept-all-defaults option.
    If you have the script locally and would like to know more about
    input options for this script, then you can run:
    ./install.sh -h
    If you would like to know more about input options for this script, refer to:
    https://github.com/oracle/oci-cli/blob/master/scripts/install/README.rst
    ******************************************************************************
Downloading Oracle Cloud Infrastructure CLI install script from https://raw.githubusercontent.com/oracle/oci-cli/v2.14.4/scripts/install/install.py to /tmp/oci_cli_install_tmp_xUxK.
######################################################################## 100.0%
Python not found on system PATH
Running install script.
python3 /tmp/oci_cli_install_tmp_xUxK
-- Verifying Python version.
-- Python version 3.5.2 okay.

===> In what directory would you like to place the install? (leave blank to use '/home/ubuntu/lib/oracle-cli'):
-- Creating directory '/home/ubuntu/lib/oracle-cli'.
-- We will install at '/home/ubuntu/lib/oracle-cli'.

===> In what directory would you like to place the 'oci' executable? (leave blank to use '/home/ubuntu/bin'):
-- Creating directory '/home/ubuntu/bin'.
-- The executable will be in '/home/ubuntu/bin'.

===> In what directory would you like to place the OCI scripts? (leave blank to use '/home/ubuntu/bin/oci-cli-scripts'):
-- Creating directory '/home/ubuntu/bin/oci-cli-scripts'.
-- The scripts will be in '/home/ubuntu/bin/oci-cli-scripts'.

===> Currently supported optional packages are: ['db (will install cx_Oracle)']
What optional CLI packages would you like to be installed (comma separated names; press enter if you don't need any optional packages)?:
-- The optional packages installed will be ''.
-- Executing: ['sudo', 'apt-get', 'update']
Hit:1 http://iad-ad-2.clouds.archive.ubuntu.com/ubuntu xenial InRelease
Hit:2 http://iad-ad-2.clouds.archive.ubuntu.com/ubuntu xenial-updates InRelease
Hit:3 http://iad-ad-2.clouds.archive.ubuntu.com/ubuntu xenial-backports InRelease
Hit:4 http://security.ubuntu.com/ubuntu xenial-security InRelease
Reading package lists... Done
-- Installing python3-venv.
-- Executing: ['sudo', 'apt-get', 'install', 'python3-venv', '-y']
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following package was automatically installed and is no longer required:
  grub-pc-bin
Use 'sudo apt autoremove' to remove it.
The following additional packages will be installed:
  python-pip-whl python3.5-venv
The following NEW packages will be installed:
  python-pip-whl python3-venv python3.5-venv
0 upgraded, 3 newly installed, 0 to remove and 16 not upgraded.
Need to get 1,119 kB of archives.
After this operation, 1,262 kB of additional disk space will be used.
Get:1 http://iad-ad-2.clouds.archive.ubuntu.com/ubuntu xenial-updates/universe amd64 python-pip-whl all 8.1.1-2ubuntu0.6 [1,112 kB]
Get:2 http://iad-ad-2.clouds.archive.ubuntu.com/ubuntu xenial-updates/universe amd64 python3.5-venv amd64 3.5.2-2ubuntu0~16.04.12 [5,996 B]
Get:3 http://iad-ad-2.clouds.archive.ubuntu.com/ubuntu xenial/universe amd64 python3-venv amd64 3.5.1-3 [1,106 B]
Fetched 1,119 kB in 0s (1,507 kB/s)
Selecting previously unselected package python-pip-whl.
(Reading database ... 62331 files and directories currently installed.)
Preparing to unpack .../python-pip-whl_8.1.1-2ubuntu0.6_all.deb ...
Unpacking python-pip-whl (8.1.1-2ubuntu0.6) ...
Selecting previously unselected package python3.5-venv.
Preparing to unpack .../python3.5-venv_3.5.2-2ubuntu0~16.04.12_amd64.deb ...
Unpacking python3.5-venv (3.5.2-2ubuntu0~16.04.12) ...
Selecting previously unselected package python3-venv.
Preparing to unpack .../python3-venv_3.5.1-3_amd64.deb ...
Unpacking python3-venv (3.5.1-3) ...
Processing triggers for man-db (2.7.5-1) ...
Setting up python-pip-whl (8.1.1-2ubuntu0.6) ...
Setting up python3.5-venv (3.5.2-2ubuntu0~16.04.12) ...
Setting up python3-venv (3.5.1-3) ...
-- Trying to use python3 venv.
-- Executing: ['/usr/bin/python3', '-m', 'venv', '/home/ubuntu/lib/oracle-cli']
-- Executing: ['/home/ubuntu/lib/oracle-cli/bin/pip', 'install', '--upgrade', 'pip']
Collecting pip
  Downloading https://files.pythonhosted.org/packages/fe/ef/60d7ba03b5c442309ef42e7d69959f73aacccd0d86008362a681c4698e83/pip-21.0.1-py3-none-any.whl (1.5MB)
    100% |████████████████████████████████| 1.5MB 1.1MB/s
Installing collected packages: pip
  Found existing installation: pip 8.1.1
    Uninstalling pip-8.1.1:
      Successfully uninstalled pip-8.1.1
Successfully installed pip-21.0.1
-- Executing: ['/home/ubuntu/lib/oracle-cli/bin/pip', 'install', '--cache-dir', '/tmp/tmp8dcrw35m', 'wheel', '--upgrade']
Traceback (most recent call last):
  File "/home/ubuntu/lib/oracle-cli/bin/pip", line 7, in <module>
    from pip._internal.cli.main import main
  File "/home/ubuntu/lib/oracle-cli/lib/python3.5/site-packages/pip/_internal/cli/main.py", line 60
    sys.stderr.write(f"ERROR: {exc}")
                                   ^
SyntaxError: invalid syntax
-- System was unable to use venv, is going to download and create virtualenv.
-- Downloading virtualenv package from https://github.com/pypa/virtualenv/archive/15.0.0.tar.gz.
-- Downloaded virtualenv package to /tmp/tmp8dcrw35m/15.0.0.tar.gz.
-- Checksum of /tmp/tmp8dcrw35m/15.0.0.tar.gz OK.
-- Extracting '/tmp/tmp8dcrw35m/15.0.0.tar.gz' to '/tmp/tmp8dcrw35m'.
-- Executing: ['/usr/bin/python3', 'virtualenv.py', '--python', '/usr/bin/python3', '/home/ubuntu/lib/oracle-cli']
Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in /home/ubuntu/lib/oracle-cli/bin/python3
Traceback (most recent call last):
  File "virtualenv.py", line 2311, in <module>
    main()
  File "virtualenv.py", line 703, in main
    symlink=options.symlink)
  File "virtualenv.py", line 916, in create_environment
    site_packages=site_packages, clear=clear, symlink=symlink))
  File "virtualenv.py", line 1209, in install_python
    shutil.copyfile(executable, py_executable)
  File "/usr/lib/python3.5/shutil.py", line 98, in copyfile
    raise SameFileError("{!r} and {!r} are the same file".format(src, dst))
shutil.SameFileError: '/usr/bin/python3' and '/home/ubuntu/lib/oracle-cli/bin/python3' are the same file
Traceback (most recent call last):
  File "/tmp/oci_cli_install_tmp_xUxK", line 254, in create_virtualenv
    upgrade_pip_wheel(tmp_dir, install_dir)
  File "/tmp/oci_cli_install_tmp_xUxK", line 236, in upgrade_pip_wheel
    exec_command(cmd)
  File "/tmp/oci_cli_install_tmp_xUxK", line 141, in exec_command
    subprocess.check_call(command_list, cwd=cwd, env=env)
  File "/usr/lib/python3.5/subprocess.py", line 581, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/home/ubuntu/lib/oracle-cli/bin/pip', 'install', '--cache-dir', '/tmp/tmp8dcrw35m', 'wheel', '--upgrade']' returned non-zero exit status 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/oci_cli_install_tmp_xUxK", line 685, in <module>
    main()
  File "/tmp/oci_cli_install_tmp_xUxK", line 637, in main
    create_virtualenv(tmp_dir, install_dir)
  File "/tmp/oci_cli_install_tmp_xUxK", line 257, in create_virtualenv
    download_and_create_virtualenv(tmp_dir, install_dir)
  File "/tmp/oci_cli_install_tmp_xUxK", line 202, in download_and_create_virtualenv
    exec_command(cmd, cwd=working_dir)
  File "/tmp/oci_cli_install_tmp_xUxK", line 141, in exec_command
    subprocess.check_call(command_list, cwd=cwd, env=env)
  File "/usr/lib/python3.5/subprocess.py", line 581, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3', 'virtualenv.py', '--python', '/usr/bin/python3', '/home/ubuntu/lib/oracle-cli']' returned non-zero exit status 1
```

**Comment by kernleee** [comment]
We discussed internally and we will not be fixing this issue. I've tried and verified installing with 3.5.3, so currently it seems like this only affects 3 patch versions of Python 3.5, 3.5.0, 3.5.1, 3.5.2.
Ubuntu 16.04 will also be officially end of life soon in April 2021. 
For these reasons we won't be adding a fix for these older Python 3.5 versions.

If you have Python <= 3.5.2 or 3.5 please make plans to upgrade to Python 3.6 or use oci-cli <= 2.19.0
